juju/juju
# Validate units and applications

The following is ongoing work to validate updating a series on an application.
Most of this PR shifts code around to ensure that we can test the updating of a series in a better way, as the application suite currently takes over 10minutes to run. This highlights the awful state that most of these tests are in. We should endeavor to fix this at every possibility, as this is one of the older facades this is going to be a slow process.

The mean of this PR will validate that an application can be updated to a given series based on either the local state for charmstore or local charms or the remote state for charmhub charms. This isn't that far off from what the machine manager facade was tasked to do but in reality, is different enough for the application facade that we have to reimplement it slightly differently.

----

We need to seriously think about how we allow a set-series for multiple user perspectives. I have an idea of acquiring a machine lock for ALL machines for a given application, but that may be too damaging. Alternatively, we could introduce an application lock and I'm unsure how far down the rabbit hole I want to go with that?

## QA steps

```sh
$ juju bootstrap lxd test
$ juju deploy ubuntu --series=bionic
$ juju deploy nrpe
$ juju add-relation nrpe ubuntu
$ juju set-series ubuntu focal
```

Check to see the series has been moved to focal. Ignore the `charmurl` and `charm-origin`
in this case, this will be solved in a future PR.

```sh
$ juju mongo
$ db.applications.find().pretty()
{
        "_id" : "dfa8a921-b699-4cf4-8a55-253aaf4754e7:ubuntu",
        "name" : "ubuntu",
        "model-uuid" : "dfa8a921-b699-4cf4-8a55-253aaf4754e7",
        "series" : "focal",
        "subordinate" : false,
        "charmurl" : "ch:amd64/bionic/ubuntu-19",
        "cs-channel" : "stable",
        "charm-origin" : {
                "source" : "charm-hub",
                "type" : "charm",
                "id" : "DksXQKAQTZfsUmBAGanZAhpoS4dpmXel",
                "hash" : "6bf4a85e0a92b6f4a5f58706c0f288c7657abda776c759da13d0409ed6656cae",
                "revision" : 19,
                "channel" : {
                        "risk" : "stable"
                },
                "platform" : {
                        "architecture" : "amd64",
                        "os" : "ubuntu",
                        "series" : "bionic"
                }
        },
        "charmmodifiedversion" : 0,
        "forcecharm" : false,
        "life" : 0,
        "unitcount" : 1,
        "relationcount" : 1,
        "minunits" : 0,
        "txn-revno" : NumberLong(5),
        "metric-credentials" : BinData(0,""),
        "exposed" : false,
        "scale" : 0,
        "passwordhash" : "",
        "txn-queue" : [
                "605b683f4407403a33b31ad6_e51cce2f"
        ]
}
{
        "_id" : "dfa8a921-b699-4cf4-8a55-253aaf4754e7:nrpe",
        "name" : "nrpe",
        "model-uuid" : "dfa8a921-b699-4cf4-8a55-253aaf4754e7",
        "series" : "focal",
        "subordinate" : true,
        "charmurl" : "ch:amd64/bionic/nrpe-70",
        "cs-channel" : "stable",
        "charm-origin" : {
                "source" : "charm-hub",
                "type" : "charm",
                "id" : "I4k7tN5SKZr4aYJ8b1GzufIDUQfievyT",
                "hash" : "1dc14c5215288cc63b19610e1596bef4f447e29ee2655b34b6d248689f19cce7",
                "revision" : 70,
                "channel" : {
                        "risk" : "stable"
                },
                "platform" : {
                        "architecture" : "amd64",
                        "os" : "ubuntu",
                        "series" : "bionic"
                }
        },
        "charmmodifiedversion" : 0,
        "forcecharm" : false,
        "life" : 0,
        "unitcount" : 1,
        "relationcount" : 1,
        "minunits" : 0,
        "txn-revno" : NumberLong(5),
        "metric-credentials" : BinData(0,""),
        "exposed" : false,
        "scale" : 0,
        "passwordhash" : "",
        "txn-queue" : [
                "605b683f4407403a33b31ad6_e51cce2f"
        ]
}
```
